shacl12-core/index.html

```diff
@@ -2289,219 +2275,179 @@ <h3>SHACL Property Paths</h3>
+				</p>
+			</section>
+			<section id="property-path-zero-or-one">
+				<h3>Zero-Or-One Paths</h3>
+				<p class="syntax" data-syntax-rule="path-zero-or-one">
+					A <dfn>zero-or-one path</dfn> is a <a>blank node</a> that is the <a>subject</a> of exactly one <a>triple</a> in <code>G</code>.
+					This triple has <code>sh:zeroOrOnePath</code> as <a>predicate</a>, and the <a>object</a> <code>v</code> is a <a>well-formed</a> SHACL property path.
+				</p>
+				<p>
+					If <code>p</code> is a <a>zero-or-one path</a> in <code>G</code>, then <code>path(p,G)</code> is a
+					SPARQL <code>ZeroOrOnePath</code> with <code>path(v,G)</code> as its <code>elt</code>.
+				</p>
+			</section>
+		</section>
+
+		<section id="node-expressions">
+			<h2>Node Expressions</h2>
+			<p>
+				This section introduces the concept of <a>node expressions</a>.
+				SHACL Core supports node expressions in the following features:
+			</p>
+			<ul>
+				<li>At <a href="#property-shapes"><code>sh:values</code> and <code>sh:defaultValue</code></a> to derive the value nodes of a property shape.</li>
+				<li>At <a href="#targetNode"><code>sh:targetNode</code></a> to dynamically compute the targets of a shape.</li>
+				<li>As parameter values of most <a href="#core-components">constraint components</a> to represent constraints that may be different depending on each focus node. <span class="todo">TODO: This change needs to be made still, see ISSUE 311.</span></li>
+				<li>At <a href="#deactivated"><code>sh:deactivated</code></a> to deactivate certain shapes under specific conditions. <span class="todo">TODO: This change needs to be made still, see ISSUE 338.</span></li>
+			</ul>
+			<p>
+				Readers who are only interested in SHACL Core can typically skip this section.
+				Given that Core only supports constant IRIs and literals as node expressions, the use cases of node expressions
+				are identical to traditional use of SHACL Core.
+			</p>
+			<div class="def">
+				<div class="def-header">TEXTUAL DEFINITION</div>
+				A <dfn>node expression</dfn> is a <a>node</a> that follows the syntax rules of exactly one <dfn>node expression function</dfn>.
+				Each <a>node expression function</a> has an <a>IRI</a> as its <dfn>function name</dfn>.
+			</div>
+			<div class="def" id="node-expression-evaluation">
+				<div class="def-header">EVALUATION OF NODE EXPRESSIONS</div>
+				The <dfn>evaluation</dfn> of a node expression is defined as a function <code>evalExpr(expr, activeGraph, scope) -> outputNodes</code>
+				where
+				<ul>
+					<li><code>expr</code> is a <a>node expression</a>.</li>
+					<li><code>activeGraph</code> is a <a>graph</a>, called the <dfn>active graph</dfn>.</li>
+					<li><code>scope</code> is a map from <a href="https://www.w3.org/TR/sparql12-query/#defn_QueryVariable">variable names</a> to individual <a>nodes</a>.
+						The value of the variable <code>focusNode</code> (if it exists) is called the <dfn>input focus node</dfn>.
+					</li>
+				</ul>
```

Comment:
It's not clearly stated what is the `activeGraph`, but I guess it means the data graph. But shouldn't an expression also have access to the shapes graph? For example, the `sh:prefixes` argument of a `sh:sparqlExpr` is a named node pointing to the prefix declaration in the shapes graph and, therefore, requires access to the shapes graph.

One could see `expr` as a reference to the node expression in the shapes graph, including the shapes graph. But that should be better emphasized.

Reply:
The activeGraph is just a variable and which graph that is doesn't matter to the algorithm. Usually it will be the data graph (of a validation engine) but I assume node expressions will be used from various other places in the future too.

I agree the shapes graph access needs to be clarified. Will have pushed a small update attempting to do so.